impeller/renderer/backend/metal/compute_pass_mtl.mm

```diff
@@ -0,0 +1,251 @@
+
+  const id<MTLComputeCommandEncoder> encoder_;
+  id<MTLComputePipelineState> pipeline_ = nullptr;
+  BufferMap buffers_;
+  TextureMap textures_;
+  SamplerMap samplers_;
+};
+
+static bool Bind(PassBindingsCache& pass,
+                 Allocator& allocator,
+                 size_t bind_index,
+                 const BufferView& view) {
+  if (!view.buffer) {
+    return false;
+  }
+
+  auto device_buffer = view.buffer->GetDeviceBuffer(allocator);
+  if (!device_buffer) {
+    return false;
+  }
+
+  auto buffer = DeviceBufferMTL::Cast(*device_buffer).GetMTLBuffer();
+  // The Metal call is a void return and we don't want to make it on nil.
+  if (!buffer) {
+    return false;
+  }
+
+  pass.SetBuffer(bind_index, view.range.offset, buffer);
+  return true;
+}
+
+static bool Bind(PassBindingsCache& pass,
+                 size_t bind_index,
+                 const Texture& texture) {
+  if (!texture.IsValid()) {
+    return false;
+  }
+
+  pass.SetTexture(bind_index, TextureMTL::Cast(texture).GetMTLTexture());
+  return true;
+}
+
+static bool Bind(PassBindingsCache& pass,
+                 size_t bind_index,
+                 const Sampler& sampler) {
+  if (!sampler.IsValid()) {
+    return false;
+  }
+
+  pass.SetSampler(bind_index, SamplerMTL::Cast(sampler).GetMTLSamplerState());
+  return true;
+}
+
+bool ComputePassMTL::EncodeCommands(
+    const std::shared_ptr<Allocator>& allocator,
+    id<MTLComputeCommandEncoder> encoder) const {
+  PassBindingsCache pass_bindings(encoder);
+
+  fml::closure pop_debug_marker = [encoder]() { [encoder popDebugGroup]; };
+  for (const auto& command : commands_) {
+    fml::ScopedCleanupClosure auto_pop_debug_marker(pop_debug_marker);
+    if (!command.label.empty()) {
+      [encoder pushDebugGroup:@(command.label.c_str())];
+    } else {
+      auto_pop_debug_marker.Release();
+    }
+
+    // TODO(dnfield): update the compute pipeline descriptor so that it can set
+    // things like workgroup size etc.
+    pass_bindings.SetComputePipelineState(
+        ComputePipelineMTL::Cast(*command.pipeline)
+            .GetMTLComputePipelineState());
+
+    for (const auto& buffer : command.bindings.buffers) {
+      if (!Bind(pass_bindings, *allocator, buffer.first,
+                buffer.second.resource)) {
+        return false;
+      }
+    }
+
+    for (const auto& texture : command.bindings.textures) {
+      if (!Bind(pass_bindings, texture.first, *texture.second.resource)) {
+        return false;
+      }
+    }
+    for (const auto& sampler : command.bindings.samplers) {
+      if (!Bind(pass_bindings, sampler.first, *sampler.second.resource)) {
+        return false;
+      }
+    }
+  }
+  // TODO(dnfield): use feature detection to support non-uniform threadgroup
+  // sizes.
+  [encoder dispatchThreadgroups:MTLSizeMake(32, 32, 1)
+          threadsPerThreadgroup:MTLSizeMake(32, 32, 1)];
+
+  // TODO(dnfield): add some kind of completion handling logic for reading
```

Comment:
IMO, this would need the addition of fences/events as a separate API that allows for resource synchronization between not only compute passes but compute and render passes as well.